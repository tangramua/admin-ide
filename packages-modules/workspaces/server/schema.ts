

/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: addWorkspace
// ====================================================

export interface addWorkspace_addWorkspace_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface addWorkspace_addWorkspace {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  icon_url: string | null;
  spec: addWorkspace_addWorkspace_spec | null;
}

export interface addWorkspace {
  addWorkspace: addWorkspace_addWorkspace | null;
}

export interface addWorkspaceVariables {
  request: IWorkspaceCreate_Input;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: removeWorkpsace
// ====================================================

export interface removeWorkpsace {
  removeWorkspace: boolean | null;
}

export interface removeWorkpsaceVariables {
  request: IWorkspaceRemove_Input;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: setEnvVariables
// ====================================================

export interface setEnvVariables {
  setEnvVariables: boolean | null;
}

export interface setEnvVariablesVariables {
  request: IWorkspaceVariables_Input;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: startWorkspace
// ====================================================

export interface startWorkspace {
  startWorkspace: boolean | null;
}

export interface startWorkspaceVariables {
  request: WorkspaceStart_Input;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: stopWorkspace
// ====================================================

export interface stopWorkspace {
  stopWorkspace: boolean | null;
}

export interface stopWorkspaceVariables {
  request: IWorkspaceStop_Input;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: updateWorkspace
// ====================================================

export interface updateWorkspace_updateWorkspace_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface updateWorkspace_updateWorkspace {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  icon_url: string | null;
  spec: updateWorkspace_updateWorkspace_spec | null;
}

export interface updateWorkspace {
  updateWorkspace: updateWorkspace_updateWorkspace | null;
}

export interface updateWorkspaceVariables {
  request: IWorkspaceUpdate_Input;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: getWorkspace
// ====================================================

export interface getWorkspace_workspace_config_ports {
  application: string | null;
}

export interface getWorkspace_workspace_config_globalVariables {
  field: string | null;
  value: string | null;
}

export interface getWorkspace_workspace_config_variables {
  field: string | null;
  value: string | null;
}

export interface getWorkspace_workspace_config {
  ports: getWorkspace_workspace_config_ports | null;
  globalVariables: (getWorkspace_workspace_config_globalVariables | null)[] | null;
  variables: (getWorkspace_workspace_config_variables | null)[] | null;
}

export interface getWorkspace_workspace_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface getWorkspace_workspace_launch {
  port: number | null;
}

export interface getWorkspace_workspace_env {
  field: string | null;
  value: string | null;
}

export interface getWorkspace_workspace_stacks {
  type: string | null;
  chartName: string;
  repository: string;
  releaseName: string;
  chartVersion: string;
  namespace: string;
  connectionId: string;
}

export interface getWorkspace_workspace_projects_source_parameters {
  branch: string | null;
}

export interface getWorkspace_workspace_projects_source {
  type: string | null;
  location: string | null;
  providers: string | null;
  parameters: getWorkspace_workspace_projects_source_parameters | null;
}

export interface getWorkspace_workspace_projects {
  _id: string;
  name: string;
  source: getWorkspace_workspace_projects_source | null;
}

export interface getWorkspace_workspace {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  namespace: string | null;
  orgId: string | null;
  icon_url: string | null;
  connectionId: string | null;
  teamId: string | null;
  config: getWorkspace_workspace_config | null;
  spec: getWorkspace_workspace_spec | null;
  launch: getWorkspace_workspace_launch | null;
  env: (getWorkspace_workspace_env | null)[] | null;
  stacks: (getWorkspace_workspace_stacks | null)[] | null;
  projects: (getWorkspace_workspace_projects | null)[] | null;
  updatedAt: string | null;
  createdAt: string | null;
}

export interface getWorkspace {
  workspace: getWorkspace_workspace | null;
}

export interface getWorkspaceVariables {
  id: string;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL query operation: getWorkspaces
// ====================================================

export interface getWorkspaces_workspaces_config_ports {
  application: string | null;
}

export interface getWorkspaces_workspaces_config_globalVariables {
  field: string | null;
  value: string | null;
}

export interface getWorkspaces_workspaces_config_variables {
  field: string | null;
  value: string | null;
}

export interface getWorkspaces_workspaces_config {
  ports: getWorkspaces_workspaces_config_ports | null;
  globalVariables: (getWorkspaces_workspaces_config_globalVariables | null)[] | null;
  variables: (getWorkspaces_workspaces_config_variables | null)[] | null;
}

export interface getWorkspaces_workspaces_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface getWorkspaces_workspaces_launch {
  port: number | null;
}

export interface getWorkspaces_workspaces_env {
  field: string | null;
  value: string | null;
}

export interface getWorkspaces_workspaces {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  icon_url: string | null;
  config: getWorkspaces_workspaces_config | null;
  spec: getWorkspaces_workspaces_spec | null;
  launch: getWorkspaces_workspaces_launch | null;
  env: (getWorkspaces_workspaces_env | null)[] | null;
  updatedAt: string | null;
  createdAt: string | null;
}

export interface getWorkspaces {
  workspaces: (getWorkspaces_workspaces | null)[] | null;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL subscription operation: subscribeToWorkspace
// ====================================================

export interface subscribeToWorkspace_subscribeToWorkspace_value_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface subscribeToWorkspace_subscribeToWorkspace_value {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  icon_url: string | null;
  spec: subscribeToWorkspace_subscribeToWorkspace_value_spec | null;
}

export interface subscribeToWorkspace_subscribeToWorkspace {
  value: subscribeToWorkspace_subscribeToWorkspace_value | null;
  mutation: WorkspaceServerEvents | null;
}

export interface subscribeToWorkspace {
  subscribeToWorkspace: subscribeToWorkspace_subscribeToWorkspace | null;
}

export interface subscribeToWorkspaceVariables {
  filter?: SubscribeWorkspaceFilter | null;
  mutations: (WorkspaceServerEvents | null)[];
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: WorkspaceDetail
// ====================================================

export interface WorkspaceDetail_config_variables {
  field: string | null;
  value: string | null;
}

export interface WorkspaceDetail_config {
  variables: (WorkspaceDetail_config_variables | null)[] | null;
}

export interface WorkspaceDetail_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface WorkspaceDetail_launch {
  port: number | null;
}

export interface WorkspaceDetail_env {
  field: string | null;
  value: string | null;
}

export interface WorkspaceDetail_projects_source_parameters {
  branch: string | null;
  httpsUrl: string | null;
  isPrivate: boolean | null;
}

export interface WorkspaceDetail_projects_source {
  type: string | null;
  language: string | null;
  location: string | null;
  providers: string | null;
  parameters: WorkspaceDetail_projects_source_parameters | null;
}

export interface WorkspaceDetail_projects {
  name: string;
  description: string | null;
  source: WorkspaceDetail_projects_source | null;
}

export interface WorkspaceDetail {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  icon_url: string | null;
  config: WorkspaceDetail_config | null;
  spec: WorkspaceDetail_spec | null;
  launch: WorkspaceDetail_launch | null;
  env: (WorkspaceDetail_env | null)[] | null;
  projects: (WorkspaceDetail_projects | null)[] | null;
  updatedAt: string | null;
  createdAt: string | null;
}


/* tslint:disable */
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL fragment: workspaceInfo
// ====================================================

export interface workspaceInfo_spec {
  cpu: number | null;
  ram: number | null;
  hdd: number | null;
}

export interface workspaceInfo {
  id: string | null;
  name: string | null;
  language: string | null;
  status: string | null;
  description: string | null;
  icon_url: string | null;
  spec: workspaceInfo_spec | null;
}

/* tslint:disable */
// This file was automatically generated and should not be edited.

//==============================================================
// START Enums and Input Objects
//==============================================================

export enum WorkspaceServerEvents {
  WORKSPACE_CREATED_EVENT = "WORKSPACE_CREATED_EVENT",
  WORKSPACE_REMOVED_EVENT = "WORKSPACE_REMOVED_EVENT",
  WORKSPACE_STARTED_EVENT = "WORKSPACE_STARTED_EVENT",
  WORKSPACE_STOPPED_EVENT = "WORKSPACE_STOPPED_EVENT",
  WORKSPACE_UPDATED_EVENT = "WORKSPACE_UPDATED_EVENT",
}

export interface WorkspaceRepository_Input {
  description: string
  name: string
  owner: string
  url: string
}

// 
export interface IWorkspaceCreate_Input {
  name: string;
  teamId?: string | null;
  repository?: WorkspaceRepository_Input;
  language?: string | null;
  spec?: IWsSpecInput | null;
  stacks?: (IStackInput | null)[] | null;
  description?: string | null;
  projects?: (IProject_Input | null)[] | null;
}

// 
export interface IWsSpecInput {
  inactivity?: number | null;
}

// 
export interface IStackInput {
  type?: string | null;
  valuesFile?: string | null;
  chartName: string;
  repository: string;
  releaseName: string;
  chartVersion: string;
  variables?: (IEnvVariableInput | null)[] | null;
}

// 
export interface IEnvVariableInput {
  field?: string | null;
  value?: string | null;
  secured?: boolean | null;
}

// 
export interface IProject_Input {
  path?: string | null;
  name: string;
  description?: string | null;
  source?: IProjectSource | null;
}

// 
export interface IProjectSource {
  type?: string | null;
  location?: string | null;
  providers?: string | null;
  parameters?: ISourceParams | null;
}

// 
export interface ISourceParams {
  branch?: string | null;
  httpsUrl?: string | null;
  isPrivate?: boolean | null;
}

// 
export interface IWorkspaceRemove_Input {
  id: string;
}

// 
export interface IWorkspaceVariables_Input {
  workspace: string;
  variables?: (IEnvVariableInput | null)[] | null;
  ports?: InputWorkspacePorts | null;
  globalVariables?: (IEnvVariableInput | null)[] | null;
}

// 
export interface InputWorkspacePorts {
  application?: string | null;
  management?: string | null;
  worker?: string | null;
}

// 
export interface WorkspaceStart_Input {
  id: string;
  requestedUserId?: string | null;
  stacks?: (IStackInput | null)[] | null;
}

// 
export interface IWorkspaceStop_Input {
  id: string;
  requestedUserId?: string | null;
  stacks?: (IStackInput | null)[] | null;
}

// 
export interface IWorkspaceUpdate_Input {
  id: string;
  name?: string | null;
  language?: string | null;
  description?: string | null;
  status?: string | null;
  icon_url?: string | null;
  os_version?: string | null;
}

// 
export interface SubscribeWorkspaceFilter {
  serverId?: string | null;
  creatorId?: string | null;
}

//==============================================================
// END Enums and Input Objects
//==============================================================