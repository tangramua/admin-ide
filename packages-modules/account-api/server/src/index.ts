
import { Feature } from '@common-stack/server-core';

export * from './constants';
export * from './interfaces';
import OrganizationFeature from './module';

export default new Feature(OrganizationFeature);
