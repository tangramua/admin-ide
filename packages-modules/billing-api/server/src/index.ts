import { interfaces } from 'inversify';
import { TaggedType } from '@common-stack/core';
import { Feature } from '@common-stack/server-core';

import { ETypes } from './constants';
import { billingStoreModule } from './store';
import { IBillingService } from './interfaces';
import { billingModule, billingModuleNats } from './container';

export * from './constants';
export * from './interfaces';

const billingServicesFunc = (container: interfaces.Container) => {
    const environment = container.get('Environment');
    if (environment === 'development') {
        return { billingService: container.get<IBillingService>(ETypes.BillingService) };
    } else {
        return { billingService: container.getNamed<IBillingService>(ETypes.BillingService, TaggedType.MICROSERVICE) };
    }
};

export default new Feature({
    createServiceFunc: billingServicesFunc,
    createContainerFunc: [billingModule, billingModuleNats, billingStoreModule],
});
