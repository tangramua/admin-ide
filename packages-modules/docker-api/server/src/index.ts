import module from './module';
import { Feature } from '@common-stack/server-core';

export * from './constants';
export default new Feature(module);

