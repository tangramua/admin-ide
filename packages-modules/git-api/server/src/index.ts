import { Feature } from '@common-stack/server-core';
import module from './module';

export * from './errors';
export * from './graphql';
export default new Feature(module);
