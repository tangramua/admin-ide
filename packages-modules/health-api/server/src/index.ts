import { Feature } from '@common-stack/server-core';
import module from './module';

export default new Feature(module);
