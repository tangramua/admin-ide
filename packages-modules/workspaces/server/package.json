{
    "name": "@adminide-stack/workspaces-server",
    "version": "0.0.1",
    "description": "Sample core for higher packages to depend on",
    "main": "lib/index.js",
    "scripts": {
        "jest": "./node_modules/.bin/jest",
        "test": "cross-env ENV_FILE=../../config/test/test.env jest",
        "test:debug": "npm test -- --runInBand",
        "test:watch": "npm test -- --watch",
        "prepublish": "npm run build",
        "watch": "npm run build:lib:watch",
        "build": "npm run build:clean && npm run build:lib",
        "build:clean": "rimraf lib",
        "build:lib": "webpack",
        "build:lib:watch": "npm run build:lib -- --watch",
        "githubschema:download": "./node_modules/.bin/apollo-codegen download-schema https://api.github.com/graphql --output githubapi-schema.json  --header 'Authorization: Bearer 9231da9aedb863f3c56329ca9d821252b247c9e2'",
        "githubschema:generate": "./node_modules/.bin/apollo-codegen generate src/browser/graphql/**/*.graphql --schema githubapi-schema.json --target typescript --output src/browser/types/github-schema.ts"
    },
    "jest": {
        "testEnvironment": "node",
        "roots": [
            "src"
        ],
        "modulePaths": [
            "node_modules"
        ],
        "moduleFileExtensions": [
            "ts",
            "tsx",
            "js",
            "json"
        ],
        "transform": {
            "\\.(ts|tsx)$": "<rootDir>/../../node_modules/ts-jest/preprocessor.js"
        },
        "testRegex": "/__tests__/.*test*\\.(ts|tsx|js)$"
    },
    "author": "CDMBase LLC",
    "license": "ISC",
    "dependencies": {
        "@adminide-stack/account": "file:../../../packages/adminide-account",
        "@adminide-stack/core": "file:../../../packages/adminide-core",
        "@adminide-stack/user-core": "file:../../../packages/user-core",
        "nanoid": "^1.0.4"
    },
    "typings": "lib/index.d.ts",
    "typescript": {
        "definition": "lib/index.d.ts"
    }
}
