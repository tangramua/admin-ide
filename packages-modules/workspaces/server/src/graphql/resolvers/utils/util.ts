import * as _ from 'lodash';
import {
    IProjectSourceType, IWorkspaceCreate_Input, IWorkspace,
    IWorkspaceMetadata, IProject_Input,
} from '@adminide-stack/core';
import { config } from '../../../config';
const generate = require('nanoid/generate');

export const getToken = (provider, meta) => {
    const list = _.get(meta, 'identity');
    const identity = _.find(list, record => record.provider.toLowerCase() === provider.toLowerCase());
    return _.get(identity, 'access_token', false);
};

export function transformWorkspace(workspace, metadata: IWorkspaceMetadata) {
    const shortId = generate('1234567890abcdefghijklmnopqrstuvwxyz', 5);

    const connectionId = generate('0123456789abcdefghijklmnopqrstuvwxyz', 15);
    const namespace = `${metadata.namespace}-${shortId}`;

    const output: any = _.transform(workspace, function (res, value, key: 'stacks' | 'projects') {
        let newStacValue, newProjectValue;
        if (key === 'stacks') {
            newStacValue = _.transform<any, any>(value, function (stackRes, stackValue, stackKey) {
                stackRes.push({ ...stackValue, releaseName: `${namespace}-${stackValue.releaseName}`, namespace, connectionId });
            });
            res = _.assign(res, { stacks: newStacValue });
        } else {
            res = _.assign(res, { [key]: value });
        }


    }, { orgId: metadata.orgId, namespace: namespace, config: { metaInfo: { domainName: config.CDE_WORKSPACE_DOMAIN } } });

    return output;
}


const getSshkey = (provider, userData) => {
    const list = _.get(userData, 'identities');
    const identity = _.find(list, record => record.provider.toLowerCase() === provider.toLowerCase());
    return _.get(identity, 'access_token', false);
};

export function transformToUserSecrets(sshData, authMetadata) {
    return {
        keys: [{
            name: 'gitsecret',
            privateKey: sshData.key,
            dir: '/etc/sshkey_git',
            containerName: ['git-server'],
        }],
        tokens: {
            GITHUB_TOKEN: getSshkey('github', authMetadata),
            GITLAB_TOKEN: getSshkey('gitlab', authMetadata),
            BITBUCKET_TOKEN: getSshkey('bitbucket', authMetadata),
        },
    };
}

const getUserParams = (provider, userData) => {
    const list = _.get(userData, 'identities');
    const identity = _.find(list, record => record.provider.toLowerCase() === provider.toLowerCase());
    return {
        username: _.get(identity, 'profileData.nickname'),
        email: _.get(identity, 'profileData.email'),
    };
};

export function transformToUserProject(projects: IProject_Input[], authMetadata) {

    const globals = {
        username: '',
        email: '',
        sshkeyPath: '/etc/sshkey_git/gitsecret',
    };

    const consolidateProject = projects.reduce((acc, cur, i) => {
        if (!!cur && (cur.source.type === IProjectSourceType.Git)) {
            const provider = cur.source.providers;
            return { ...acc, [provider]: { ...getUserParams(provider, authMetadata) } };
        }
        return acc;
    }, {});
    return { ...consolidateProject, globals };
}
