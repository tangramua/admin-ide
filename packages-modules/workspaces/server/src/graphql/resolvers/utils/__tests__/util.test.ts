

import 'jest';
import { transformWorkspace } from '../util';
import * as _ from 'lodash';
import { logger } from '@cdm-logger/server';

describe('transformWorskapce utlity test', () => {

    it('test', () => {
        const workspace = {
            'name': 'dfsfsdf',
            'stacks': [
                {
                    'type': 'MONOCULAR_STACK',
                    'chartName': 'aerospike',
                    'repository': 'stable',
                    'releaseName': 'sdfsafdsf',
                    'chartVersion': '0.1.7',
                    'variables': [
                        {

                        },
                    ],
                },
            ],
            'projects': [
                {
                    'name': 'sdarewr',
                    'source': {
                        'type': 'BLANK',
                        'location': '',
                        'providers': 'NONE',
                        'parameters': {
                            'branch': '',
                        },
                    },
                },
                {
                    'name': 'jLJfwer',
                    'source': {
                        'type': 'GIT',
                        'location': '',
                        'providers': 'GITHUB',
                        'parameters': {
                            'branch': '',
                        },
                    },
                },
            ],
        };
        const metadata: any = {
            'namespace': 'stackflow1',
            'identity': [
                {
                    'provider': 'google-oauth2', 'access_token': 'googleccesssstoookkkkeeeen',
                    'expires_in': 3600, 'user_id': '100879977420', 'connection': 'google-oauth2', 'isSocial': true,
                }, {
                    'provider': 'github', 'access_token': 'gitesssstoookkkkeeeen',
                    'expires_in': 3600, 'user_id': '2342342', 'connection': 'github', 'isSocial': true,
                },
            ],
            'orgId': '5b31afefcd52937450187a2d',
        };
        const output = transformWorkspace(workspace, metadata);
        expect(output).toMatchSnapshot();
        logger.info('output [%o]', output);

    });
});
