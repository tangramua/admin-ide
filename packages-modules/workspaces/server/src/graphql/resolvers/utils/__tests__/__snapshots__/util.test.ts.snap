// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformWorskapce utlity test test 1`] = `
Object {
  "name": "dfsfsdf",
  "namespace": "stackflow1-59ed98cca4",
  "orgId": "5b31afefcd52937450187a2d",
  "projects": Array [
    Object {
      "name": "sdarewr",
      "source": Object {
        "location": "",
        "parameters": Array [
          Object {
            "branch": "",
          },
        ],
        "providers": "NONE",
        "type": "BLANK",
      },
    },
    Object {
      "access_token": "gitesssstoookkkkeeeen",
      "name": "jLJfwer",
      "source": Object {
        "location": "",
        "parameters": Array [
          Object {
            "branch": "",
          },
        ],
        "providers": "GITHUB",
        "type": "GIT",
      },
    },
  ],
  "stacks": Array [
    Object {
      "chartName": "aerospike",
      "chartVersion": "0.1.7",
      "connectionId": "iyzfRgRdZwT0ODghjYQgT",
      "namespace": "stackflow1-59ed98cca4",
      "releaseName": "stackflow1-59ed98cca4-sdfsafdsf",
      "repository": "stable",
      "type": "MONOCULAR_STACK",
      "variables": Array [
        Object {},
      ],
    },
  ],
}
`;
